PurchasesCoreSwift/StoreKitExtensions/SKPaymentTransaction+Extensions.swift

```diff
@@ -0,0 +1,38 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  SKPaymentTransaction+Extensions.swift
+//
+//  Created by Juanpe Catalán on 2/8/21.
+
+import StoreKit
+
+// todo: make internal after migration
+public extension SKPaymentTransaction {
+
+    /// Although neither `payment` nor `productIdentier` is optional, we must continue
+    /// checking the nullability to likely fix issue https://github.com/RevenueCat/purchases-ios/issues/279
+    @objc var rc_productIdentifier: String? {
+        let maybePayment: SKPayment? = payment
+
+        guard maybePayment != nil else {
```

Comment:
we're not doing the usual `guard let payment = payment` here to keep an explicit comparison `!= nil`, right?

Reply:
You're right. TBH, this code is tricky 😅 since we are storing an optional reference to "cheat" to the compiler. The problem is that we want the `productIdentifier`, which is not optional, so if we unwrap the payment reference, we must create an optional reference of the product identifier. However, if we keep the payment reference as optional, we can use it to validate whether the product identifier is nil too.

I don't know if I'm explaining myself 😓...anyway...we could do it, but I'm not sure if we have more benefits than having the current one.

The code will look like this:
```swift
@objc var rc_productIdentifier: String? {
    let maybePayment: SKPayment? = payment

    guard let anyPayment = maybePayment else {
        Logger.appleWarning(Strings.purchase.skpayment_missing_from_skpaymenttransaction)
        return nil
    }
    
    let maybeProductIdentifier: String? = anyPayment.productIdentifier

    guard let productIdentifier = maybeProductIdentifier,
          !productIdentifier.isEmpty else {
        Logger.appleWarning(Strings.purchase.skpayment_missing_product_identifier)
        return nil
    }

    return productIdentifier
}
```

WDYT?

Reply:
🤔 I think the current approach is fine, and we do have comments that explain it, so it should be fine. If any devs try to refactor into a `guard let`, they'll run into a compiler issue right away, and the motivation is explained in the comments, so I don't think it's worth going crazy over it. Just wanted to double-check that this was the case.

Reply:
Exactly, this is the point 🙂 I don't know if the opinion of another developer who doesn't have the whole context could be useful to be sure that the code is understandable or if we had better modify it 🤔

Reply:
I do not have the whole context and I actually find it slightly confusing. Specifically because we don't unwrap `payment` but we go and use `maybePayment` later in the file. I think we should remove `let maybePayment: SKPayment? = payment` and use `guard let payment = payment else { ` instead of `guard let anyPayment = maybePayment else {`

Reply:
I understand your point, maybe we could update the comment to explain better what this getter does. The problem is that this code `guard let payment = payment` doesn't compile because `payment` is not optional. So? Why make it more complex than it seems? 😅 

To give you more context on this issue, we are extending the `SKPaymentTransaction` class by adding a shortcut to get the product identifier. This should be something like this: 
```swift
extension SKPaymentTransaction {
    var productIdentifier: String {
        payment.productIdentifier
    }
}
```

But considering issue #279, sometimes `payment` and `productIdentifier` can be `nil`, in this case, we have to treat them as nullable. So the approach I took was to create an optional reference, then the compiler would allow us to check for nullability.

Do you know another way to solve this problem easier? 🤔

Reply:
AHhh, right, ok. I'm a big fan of comments, and I think that would really help. @aboedo explained some the context here around incorrect nullability, so this makes more sense now. So, I'm ok with this + a comment 😄

Reply:
Hey @Juanpe, caught up on this thread and 🤯 . Thanks for taking on a tricky one...

I tried something _slightly_ different out that seems to work:

```suggestion
        guard let maybePayment = payment as SKPayment? else {
            Logger.appleWarning(Strings.purchase.skpayment_missing_from_skpaymenttransaction)
            return nil
        }
        guard let productIdentifier = maybePayment.productIdentifier as String?,
              !productIdentifier.isEmpty else {
            Logger.appleWarning(Strings.purchase.skpayment_missing_product_identifier)
            return nil
        }
```

It's basically the same as your implementation, with the first `maybePayment` bit squashed down. I think it just makes it a little more explicit that `payment` really is of type `SKPayment?` by explicitly casting it, and then doing the same for `productIdentifier`. What do you think?

Reply:
Hi @beylmk! Thanks for the suggestion, I think it's a tiny change but it reduces the logic, so I will definitely include it. I just have a question related to the swift style guide. It's totally minor, but does it make sense to use `maybePayment` yet? I mean, the prefix `maybe` is used to identify optional `vars`, but if the unwrapping operation is used as the guard condition, the var is not optional. Do you know what I mean? 
 
Some suggestions could be...
```swift
guard let payment = payment as? SKPayment else { ... }

// or

guard let anyPayment = payment as? SKPayment else { ... }

// or

guard let skPayment = payment as? SKPayment else { ... }
```

WDYT? the var should be still called `maybePayment` or we should use another one?  🤔

Reply:
🤔 intersting. My two cents is that: 
```swift
guard let maybePayment = payment as SKPayment? else {
```
indicates that `payment` can safely be interpreted as `SKPayment?` (which makes sense), and we're checking for `nil` with `guard let`, while
```swift
guard let payment = payment as? SKPayment else { ... }
```
kinda reads like we're expecting that `payment` could be of an entirely different type (i.e.: string), and that's what we're checking for. 

I don't have a preference for `maybePayment` vs `payment` in this particular context, but if the cast works correctly with ` as SKPayment?`, I'd go with that. 

That said, I think as long as the _why_ is explained nicely with a comment, either would work. 

@Juanpe I'm happy to go with what you like best here. Thoughts?

Reply:
yes, you convinced me, the second option might not be completely clear. I prefer to use `maybe` so as to indicating that the `payment` can be `nil`.

The current code uses `maybe` and I added a comment, do you think this is enough or we should add a comment related to this thread?

Reply:
I think the comment is clear enough, and the compiler and tests won't allow us to break this by refactoring it incorrectly, so we're good to go!